pyfar/classes/filter.py

```diff
@@ -537,27 +562,124 @@ def init_state(self, cshape, state='zeros'):
-        infinitely long impulse response of the filter.
+        Compute the impulse response of the filter.
 
         Parameters
         ----------
-        n_samples : int
-            Length in samples for which the impulse response is computed.
+        n_samples : int, optional
+            Length in samples for which the impulse response is computed. The
+            default is ``None`` in which case ``n_samples`` is computed as the
+            maximum value returned by :py:func:`~impulse_response_length`.
+            A warning is returned if ``n_samples`` is too short to approximate
+            the theoretically infinitely long impulse response.
 
         Returns
         -------
         impulse_response : Signal
-            The impulse response of the filter.
+            The impulse response of the filter of with a ``cshape = (C, )``
+            where `C` denotes the number of channels of the filter.
         """
         return super().impulse_response(n_samples)
 
+    def impulse_response_length(self, tolerance=5e-5):
+        """
+        Get the estimated length of filter impulse response in samples.
+
+        The length is estimated from the positions of the poles of the filter.
+        The closer the poles are to the unit circle, the longer the estimated
+        length.
+
+        Parameters
+        ----------
+        tolerance : float, optional
+            Tolerance for the accuracy. Smaller tolerances will result in
+            larger impulse response lengths. The default is ``5e-5``.
+
+        Returns
+        -------
+        impulse_response_length : array
+            An integer array of shape (C, ) containing the length in samples
+            where `C` denotes the number of channels of the filter.
+        """
+
+        channels = self.coefficients.shape[0]
+        estimated_length = np.zeros(channels)
+
+        for channel in range(channels):
+
+            b = self.coefficients[channel, 0]
+            a = self.coefficients[channel, 1]
+
+            # this is an FIR filter
+            if self.order == 0 or np.all(a[1:] == 0):
+                estimated_length[channel] = np.max(np.nonzero(b)) + 1
+                continue
+
+            # This is an IIR filter. Delay of non-recursive part
+            estimated_length[channel] = np.min(np.nonzero(b))
+
+            # poles of the transfer function
+            poles = np.roots(a)
+
+            if np.any(np.abs(poles) > 1.0001):
+                # This is an unstable filter. The closer the outmost pole is to
+                # the unit circle, the longer the impulse response.
+                estimated_length[channel] += \
+                    6 / np.log10(np.max(np.abs(poles[np.abs(poles) > 1])))
+            else:
+                # This is a stable filter. Minimum height is 1e-5 of original
+                # amplitude
+                idx = np.abs(poles - 1) < 1e-5
+                poles[idx] = -poles[idx]
```

Comment:
Why are those poles reflected wrt to the origin?
It this supposed to bring the unstable poles into the unit circle? If yes, it can be done like this:
`poles[idx] = 1/np.abs(poles[idx]) * np.exp(1j*np.angle(poles[idx]))`